File: Homework/Manh/lesson-1/SortArrayByParity.java
class Solution {
    
    public int[] sortArrayByParity(int[] nums) {
        int n = nums.length;
        int i=0;
        int j=n-1;

        while (i<j){
            while(i<j && nums[i] % 2 ==0) i++;
            while(j>i && nums[j] % 2 !=0) j--;

            if(i>=j) return nums;
            else swap(i,j,nums);
        }

        return nums;
    }
    
    private void swap(int x, int y, int[] nums){
        if(x==y) return;
        nums[x] = nums[x] + nums[y];
        nums[y] = nums[x] - nums[y];
        nums[x] = nums[x] - nums[y];
    }
}

File: Homework/Khang_Nguyen/sort-array-by-parity.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
 function swap(start, end, arr){
    const temp = arr[start]
    arr[start] = arr[end]
    arr[end] = temp
}

var sortArrayByParity = function(nums) {
    const len = nums.length
    let leftMostOdd = -1
    for (let i = 0; i < len; i++) {
        if (nums[i] % 2 !== 0) {
            if (leftMostOdd === -1) {
                leftMostOdd = i
            }
        } else {
            if (leftMostOdd !== -1) {
                swap(leftMostOdd, i, nums)
                leftMostOdd++                
            }
        }
    }
    return nums
};

// Time: O(n)
// Space: O(1)